### PR TITLE
Enhanced Delta Printer Calibration

### PR DESCRIPTION
**Pull request by KevinOConnor**
This is a request for help testing a new "enhanced delta calibration" system I've been working on.  The goal is to find delta parameters such that the bed appears level and printed objects have correct XY dimensions.

Testing requires updating to a development branch - ssh into the Raspberry Pi and run:
```
cd ~/klipper ; git fetch ; git checkout origin/work-delta-calibrate-20180828 ; sudo service klipper restart
```

Then follow the guide at: https://github.com/KevinOConnor/klipper/blob/work-delta-calibrate-20180828/docs/Delta_Calibrate.md

I'm interested in feedback (success or failure).  Please provide that feedback by adding a new comment to this github pull request.  Please also attach the klipper.log file from the session the calibration was performed on.  (See: https://github.com/KevinOConnor/klipper/blob/master/docs/Contact.md for details on obtaining and attaching the log file.)

-Kevin

### CODE REVIEW

File: klippy/extras/delta_calibrate.py [hunk 1]
```diff
@@ -6,11 +6,107 @@
...
+# How much to prefer a distance measurement over a height measurement
+MEASURE_WEIGHT = 0.5
+
+# Convert distance measurements made on the calibration object to
+# 3-tuples of (actual_distance, stable_position1, stable_position2)
+def measurements_to_distances(measured_params, delta_params):
+    # Extract params
+    mp = measured_params
+    dp = delta_params
+    scale = mp['SCALE']
```
Comment (ysoldak) [inline klippy/extras/delta_calibrate.py]: Here scale becomes a list and fails later be multiplied by MeasureRidgeRadius on line 77, resulting in exception

**Comment by ysoldak** [comment]
My calibration log.
Hit an exception, see line comment above.
Fixed it by replacing that line 65 with `scale = mp['SCALE'][0]` and here we have:
[delta_calibrate_extended_klippy.log](https://github.com/KevinOConnor/klipper/files/2359068/delta_calibrate_extended_klippy.log)

Going to print with new params now and see how it flies.

**Comment by ysoldak** [comment]
This SAVE_CONFIG command is confusing. 

It's not clear from the doc whether it is enough to SAVE_CONFIG after extended calibration and go printing directly after having all new parameters applied for you. Or you indeed shall go and manually update your `printer.cfg` (I assume that, sorry my Octoprint swallowed all the logs already by the time I went there to check if there were any instructions printed out.)

If SAVE_CONFIG saves only calibration reference data, and this does not affect printing, I'd propose to rename the command to something like SAVE_CALIBRATION_CONFIG or SAVE_CALIBRATION_REFERENCE.

**Comment by ysoldak** [comment]
Results
- Center distances in range: 49.3-49.5, for expected 49.3
- Outer distances in range: 49.1-49.4, for expected 49.3

This is a huge success, actually.
Never had my print so precise, and probably it's just impossible to have it better with 0.4 nozzle anyway.
I've printed my object, remix of yours with center-to-center distances 45.5mm and 3.8mm bolt heads, giving expected 49.3mm.

Arm lengths calculated are like 2mm longer than measured (122 vs 120mm), but who cares if it prints flat and accurate.

**Comment by KevinOConnor** [comment]
Okay - thanks.  I've fixed the scale bug.

On this development branch, both the DELTA_CALIBRATE and DELTA_ANALYZE commands will update the current session with the new delta parameters.  The SAVE_CONFIG command is needed to store those parameters to the printer.cfg file.  So, a SAVE_CONFIG command is required prior to the next RESTART command.  Going forward, the idea is to use SAVE_CONFIG for PID_CALIBRATE, BED_TILT_CALIBRATE, BED_MESH_CALIBRATE, etc.

**Comment by ysoldak** [comment]
Sorry, off topic probably, still is related to the calibration process.

>On this development branch, both the DELTA_CALIBRATE and DELTA_ANALYZE commands will update the current session with the new delta parameters. The SAVE_CONFIG command is needed to store those parameters to the printer.cfg file. So, a SAVE_CONFIG command is required prior to the next RESTART command. Going forward, the idea is to use SAVE_CONFIG for PID_CALIBRATE, BED_TILT_CALIBRATE, BED_MESH_CALIBRATE, etc.

OK, so here is the confusion.
I look into my `printer.cfg`, after restart and all, fresh power on. And see `arm_length` for `[stepper_a]`. Shall I trust this value or it gets adjusted by autogenerated `[delta_calibrate]` block?
If I change `[stepper_a]/position_endstop` value manually, will it be overridden by that autogenerated block? SAVE_CONFIG pushes me thinking in that direction.

If nothing gets overridden and autogenerated block serves just as an input for future calibrations, then this means I can't print directly after calibration is done and expect prints be better, I still have to go and update my config manually. Then SAVE_CONFIG name is misleading and name like SAVE_CALIBRATION would be better.

That said, the calibration works perfectly for me, just a bit of feedback on the commands.

**Comment by KevinOConnor** [comment]
The `[delta_calibrate]` that SAVE_CONFIG writes to printer.cfg overrides the other values.  Yeah, this may not be the best way to implement this - certainly the docs need to be updated so the user knows that the SAVE_CONFIG block contains the settings that are actually used.

Your feedback is appreciated.
-Kevin

**Comment by ysoldak** [comment]
> The [delta_calibrate] that SAVE_CONFIG writes to printer.cfg overrides the other values.

So, it just happened. I've changed the nozzle and been pulling my hear off seeing my endstop adjustments have no effect. Basically, I've tried up 0.1mm on all towers. That `[delta_calibrate]` block in the end of the file was overriding my edits, obviously.

An alternative could be load that `[delta_calibrate]` block only before next calibration? Have a flag/parameter to SAVE_CONFIG to let it change values in `[stepper_*]`  and `[printer]` blocks?

**Comment by KevinOConnor** [comment]
@ysoldak - I agree that SAVE_CONFIG is too confusing; I don't plan to commit SAVE_CONFIG to the master branch as is.  I don't think it significantly impacts enhanced delta calibration though, so I'm still interested in seeing others delta calibration results.

-Kevin

**Comment by Avalonnw** [comment]
I tried this and the result is great so far! Even bed levelling is not needed now by the looks of it!

**Comment by maketo3D** [comment]
After trying it, I can say that it works really well.
I made 3 iterations, in the first attempt I had a maximum error of 1.22mm higher than the model measure, in the last I had a maximum error of 0.05mm. I had an error of more than 3mm in the arms with respect to the measurement reported by the manufacturer of the printer.

**Comment by KevinOConnor** [comment]
@maketo3D - Thanks.  Can you attach the klipper log files from the calibration runs?  If not, can you post the measurements you used on each run?

It seems that arm length differences are a common theme with enhanced calibration.  The current thinking is that small differences in arm length can cause a tilt of the effector (a tilt that changes depending on the XY location of the head) and the calibration code is able to account for some of that tilt by using an adjusted arm length.  (So, the recommended arm lengths may be better even though they are different than the actual arm lengths.)

-Kevin

**Comment by maketo3D** [comment]
> @maketo3D - Thanks. Can you attach the klipper log files from the calibration runs? If not, can you post the measurements you used on each run?
> 
> It seems that arm length differences are a common theme with enhanced calibration. The current thinking is that small differences in arm length can cause a tilt of the effector (a tilt that changes depending on the XY location of the head) and the calibration code is able to account for some of that tilt by using an adjusted arm length. (So, the recommended arm lengths may be better even though they are different than the actual arm lengths.)
> 
> -Kevin

I just looked at the log and it shows a rollover with the last impression I made after the calibration.
[klippy.log](https://github.com/KevinOConnor/klipper/files/2378817/klippy.log)
But I have a document where I write down the commands of each calibration and the output with the results, I hope they will help.
[Calibration measures and results.txt](https://github.com/KevinOConnor/klipper/files/2378818/Calibration.measures.and.results.txt)

**Comment by ruevs** [comment]
Finally a delta calibration that makes sense!
I tried on my Anycubic Kossel (I merged the work-delta-calibrate-20180828 branch into master locally) and got an error:
`Internal error on command:"DELTA_ANALYZE"
`
The log is attached
[klippy.log](https://github.com/KevinOConnor/klipper/files/2386655/klippy.log)
Probably I should try with a clean work-delta-calibrate-20180828 branch?

**Comment by KevinOConnor** [comment]
@ruevs - you've found a bug.  For now, work around it by explicitly sending a `DELTA_ANALYZE SCALE=1.00` command.  (You should be able to issue RESTART and then reenter your distance measurements; just include the scale as well.)

-Kevin

**Comment by ruevs** [comment]
Thanks for the lightning fast reply :-) It works with a scale provided. The new log is attached.
[klippy.zip](https://github.com/KevinOConnor/klipper/files/2386711/klippy.zip)
I have no time left to test the calibration results now, but will post again when I do.

**Comment by KevinOConnor** [comment]
I have rebased this development branch and made some significant changes to how SAVE_CONFIG works.  (The DELTA_CALIBRATE and DELTA_ANALYZE commands are basically the same, but there are a couple of bug fixes.)  It is hoped that the new SAVE_CONFIG will be more transparent (if an existing config field is overwritten by SAVE_CONFIG then it will now be commented out when the printer config is rewritten, and if the user later adds a field to the printer config in the main section then it will have precedence over fields in the SAVE_CONFIG section).

If you've already run the enhanced calibration and you do not want to rerun it, then you will need to manually update your config file - move the radius, angle_[abc], arm_[abc], and endstop_[abc] in the SAVE_CONFIG block to their appropriate sections and fields in the main config section.

cd ~/klipper ; git fetch ; git checkout origin/work-delta-calibrate-20180828 ; sudo service klipper restart

-Kevin

**Comment by ysoldak** [comment]
> I have rebased this development branch and made some significant changes to how SAVE_CONFIG works.

Thanks for working on this. Is there any place to discuss save/update config ideas and approaches?

**Comment by KevinOConnor** [comment]
FYI, if there is no further comments on this, I'll merge early next week.

@ysoldak - I've opened PR #666 for discussions on config save/restore.  The mailing list is also always an option.

-Kevin

**Pull request merged**